scripts/usb_scan.py

```diff
@@ -0,0 +1,545 @@
+                continue
+            storage = UsbStorage(d.sys_name, d.sys_path)
+            if storage.is_initialized():
+                ret.append(storage)
+            else:
+                Log().debug("ignore usb storage:" + str(storage))
+        return ret
+
+
+class Policy(object):
+    __metaclass__ = Singleton
+
+    _PATH = "/etc/xensource/usb-policy.conf"
+
+    _CLASS = "class"
+    _SUBCLASS = "subclass"
+    _PROTOCOL = "prot"
+    _ID_VENDOR = "vid"
+    _ID_PRODUCT = "pid"
+    _BCD_DEVICE = "rel"
+
+    # usb_descriptor_attr_le16(idVendor, "%04x\n");
+    # usb_descriptor_attr_le16(idProduct, "%04x\n");
+    # usb_descriptor_attr_le16(bcdDevice, "%04x\n");
+
+    # key in policy <--> key in usb device
+    _KEY_MAP_DEVICE = {_ID_VENDOR: UsbDevice._ID_VENDOR, _ID_PRODUCT: UsbDevice._ID_PRODUCT,
+                       _BCD_DEVICE: UsbDevice._BCD_DEVICE}
+
+    # usb_intf_attr(bInterfaceClass, "%02x\n");
+    # usb_intf_attr(bInterfaceSubClass, "%02x\n");
+    # usb_intf_attr(bInterfaceProtocol, "%02x\n");
+
+    # key in policy <--> key in usb interface
+    _KEY_MAP_INTERFACE = {_CLASS: UsbInterface._CLASS, _SUBCLASS: UsbInterface._SUB_CLASS,
+                          _PROTOCOL: UsbInterface._PROTOCOL}
+
+    _PAT_KEY = r"\s*({}|{}|{}|{}|{}|{})\s*".format(_CLASS, _SUBCLASS, _PROTOCOL, _ID_VENDOR, _ID_PRODUCT,
+                                                   _BCD_DEVICE)
+    _PAT_VAL = r"\s*([0-9a-fA-F]+|\"[0-9a-fA-F]+\")"
+    _PATTERN = _PAT_KEY + r"=" + _PAT_VAL
+
+    _ALLOW = "allow"
+
+    def __init__(self):
+        self.rule_list = []
+        try:
+            # may raise IOError
+            with open(self._PATH, "r") as f:
+                Log().debug("=== policy file begin")
+                for line in f:
+                    Log().debug(line[0:len(line) - 1])
+                    self.parse_line(line)
+                Log().debug("=== policy file end")
+        except IOError:
+            # without policy file, no device will be allowed
+            Log().debug(traceback.format_exc())
+        Log().debug("policy list: " + str(self.rule_list))
+
+    def parse_line(self, line):
+        # 1. remove comments
+        # ^([^#]*)(#.*)?$
+        i = line.find("#")
+        if i != -1:
+            line = line[0:i]
+
+        # 2. split action and match field
+        # ^\s*(ALLOW|DENY)\s*:\s*([^:]*)$
+        if ":" not in line:
+            return
+        parts = line.split(":")
+        n = len(parts)
+        if n == 0 or n > 2:
+            return
+
+        # 3. parse action
+        # \s*(ALLOW|DENY)\s*
+        rule = {}
+        action = parts[0].strip()
+        if action == "ALLOW":
+            rule[self._ALLOW] = True
+        elif action == "DENY":
+            rule[self._ALLOW] = False
+        else:
+            return
+
+        # 4. parse key=value pairs
+        # pattern = r"\s*(class|subclass|prot|vid|pid|rel)\s*=\s*([0-9a-fA-F]+|\"[0-9a-fA-F]+\")"
```

Comment:
Rather than defining a new grammar why not just use JSON?

Reply:
The policy file is from citrix receiver team,
we just reuse its grammar.

Reply:
But what is the benefit of using this format? It is unlikely that we can re-use their data file so doing so just complicates the implementation.

Reply:
I guess it's simple, it might be easier for use to edit it.
Actually, we are using exactly the same policy file,
you can find the sample file in design page:
https://info.citrite.net/display/xenserver/Design+for+REQ-158%3A+USB+Dongles+pass-through#DesignforREQ-158:USBDonglespass-through-Dom0

Reply:
Interestingly, our rules seem to actually fail their spec: https://support.citrix.com/article/CTX132716
They do not seem to allow quotes in the pattern.

Reply:
TODO, to remove quotes from grammar.

Reply:
thanks for the spec url, I had only a sample policy file before.

Reply:
It turns out when we pasted the file content to the wiki page, the quotes was automatically added.